File: leetcode/implement_trie/implement_trie.java
package implement_trie;

class Node {
    Node[] childs;
    boolean isEnd;

    Node() {
        childs = new Node[26];
        isEnd = false;
    }
}

class Trie {
    Node head;

    public Trie() {
        head = new Node();
    }
    
    public void insert(String word) {
        Node currNode = head;
        for(char ch : word.toCharArray()) {
            int idx = ch-'a';
            if(currNode.childs[idx] == null) currNode.childs[idx] = new Node();
            currNode = currNode.childs[idx];
        }
        currNode.isEnd = true;
    }
    
    public boolean search(String word) {
        Node currNode = head;
        for(char ch : word.toCharArray()) {
            int idx = ch-'a';
            if(currNode.childs[idx] == null) return false;
            currNode = currNode.childs[idx];
        }
        return currNode.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        Node currNode = head;
        for(char ch : prefix.toCharArray()) {
            int idx = ch-'a';
            if(currNode.childs[idx] == null) return false;
            currNode = currNode.childs[idx];
        }
        return true;
        
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: leetcode/implement_trie/implement_trie.cpp
/*
 * Time Complexity:
 *   - insert: O(L), L is the length of the word.
 *   - search: O(L)
 *   - startsWith: O(L)
 *
 * Space Complexity: O(N * L)
 *   - N is the number of words, L is the average length.
 */

#include <string>

using std::string;

class Node {
public:
    Node* childs[26];
    bool isEnd;

    Node() {
        isEnd = false;
        for (auto &a : childs) a = nullptr;
    }    
};

class Trie {
private:
    Node* head;

public:
    Trie() {
        head = new Node();
    }
    
    void insert(string word) {
        Node* currNode = head;
        for(char ch : word) {
            int idx = ch-'a';
            if(!currNode->childs[idx]) currNode->childs[idx] = new Node();
            currNode = currNode->childs[idx];
        }
        currNode->isEnd = true;
    }
    
    bool search(string word, bool isPrefix = false) {
        Node* currNode = head;
        for(char ch : word) {
            int idx = ch-'a';
            if(!currNode->childs[idx]) return false;
            currNode = currNode->childs[idx];
        }
        return currNode->isEnd || isPrefix;
    }
    
    bool startsWith(string prefix) {
        return search(prefix, true);
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

int main() {
    // Example usage:
    Trie* trie = new Trie();
    trie->insert("apple");
    trie->insert("app");
    bool searchResult = trie->search("apple"); // returns true
    bool startsWithResult = trie->startsWith("app"); // returns true
    bool searchResult2 = trie->search("app"); // returns false
    delete trie; // Clean up memory
}